### ISSUE TITLE
10 Popular Machine Learning Algorithms In A Nutshell

### ISSUE THREAD

**Issue by occoder** [occoder/common_sql#583]
Understanding what Artificial Intelligence is and learning how Machine Learning and Deep Learning power it, are overwhelming experiences. In machine learning, there is something called the “No Free Lunch” theorems which states that no one machine learning algorithm works best for every problem, and it’s particularly relevant for supervised learning.

For example, you can’t say that decision trees are always better than neural networks or vice-versa. There are various factors at play, such as the size and structure of your dataset. As a result, you should try many different algorithms for your problem, while using a hold-out “test set” of data to decide performance and select the winning algorithm. 

So, In this article, we've decided to share in-depth articles, infographics, video tutorials, and python codes of some of the most common and popular machine learning algorithms. Machine learning beginners, intermediates, advanced enthusiasts as well as other learners who are curious to know about various types of machine learning algorithms and how they are useful in real world.  

Machine Learning Algorithms covered in this guide are

-    Simple Linear Regression
-   Logistic Regression
-   Decision Tree
-    Random Forest
-    SVM (Support Vector Machines)
-    KNN (k-Nearest Neighbors)
-    K-Means Clustering
-    Hierarchical Clustering
-    Multiple Linear Regression
-    Neural Networks

Note: Other widely used algorithms like Naive Bayes Classifier, XgBoost, Principal Component Analysis (Dimensionality Reduction), Perceptron, T-Distributed Stochastic Neighbor Embedding and Q Learning will be added soon.

**Linear Regression**

Initially developed in statistics to study the relationship between input and output numerical variables, it was adopted by the machine learning community to make predictions based on the linear regression equation. Linear regression can be used to find the general price trend of a stock over a period of time. This helps us understand if the price movement is positive or negative.

Generally, Linear Regression is divided into two types: Simple linear regression and Multiple linear regression

**👉 Simple Linear Regression**

Simple linear regression is a statistical method that allows us to summarize and study relationships between two continuous (quantitative) variables. It is probably one of the most popular and well-inferential algorithms in statistics and machine learning. One variable is considered as an explanatory variable and the other one as a dependent variable. 

[![](https://github.com/wiki-360/resources/raw/main/Machine%20Learning%20Algorithms%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Machine%20Learning%20Algorithms%20Explained.jpg)

**👉 Multiple Linear Regression**

Multiple linear regression (MLR), also known simply as multiple regression, is a machine learning algorithm that uses several explanatory variables to predict the outcome of a response variable. In reality, multiple regression is the extension of ordinary least-squares (OLS) regression because it includes more than one explanatory variable. 

[![](https://github.com/wiki-360/resources/raw/main/Linear%20Regression%20Machine%20Learning%20Algorithm%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Linear%20Regression%20Machine%20Learning%20Algorithm%20Explained.jpg)

**👉 Logistic Regression**

Logistic regression is part of a family of machine learning algorithms called classification algorithms.  It is the go-to method for binary classification problems. The algorithm is used to predict the probability of occurrence of an event by fitting data to a logit function. Hence, it is also known as logit regression. Customer churn, spam email, website or ad click predictions are some examples of the areas where logistic regression offers a powerful solution.

[![](https://github.com/wiki-360/resources/raw/main/Logistic%20Regression%20Algorithm%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Logistic%20Regression%20Algorithm%20Explained.jpg)

**👉 Decision Trees**

A decision tree is a flow-chart-like tree structure that uses a branching method to illustrate every possible outcome of a decision. Each node within the tree represents a test on a specific variable – and each branch is the outcome of that test. Some of the most popular decision trees algorithms are Classification and Regression Tree (CART), Iterative Dichotomiser 3 (ID3), C4.5 and C5.0, Chi-squared Automatic Interaction Detection (CHAID), Decision Stump, and more.

[![](https://github.com/wiki-360/resources/raw/main/Decision%20Tree%20Machine%20Learning%20Algorithm%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Decision%20Tree%20Machine%20Learning%20Algorithm%20Explained.jpg)

**👉 Random Forest**

The random forest algorithm establishes the outcome based on the predictions of the decision trees. It predicts by taking the average or mean of the output from various trees. Increasing the number of trees increases the precision of the outcome. A random forest eradicates the limitations of a decision tree algorithm. It reduces the overfitting of datasets and increases precision. 

[![](https://github.com/wiki-360/resources/raw/main/Random%20Forest%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Random%20Forest%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)

**👉 Support Vector Machine**

A Support Vector Machine (SVM) is a discriminative classifier formally defined by a separating hyperplane. In other words, given labeled training data (supervised learning), the algorithm outputs an optimal hyperplane which categorizes new examples. In two dimensional space this hyperplane is a line dividing a plane in two parts where in each class lay in either side. 

[![](https://github.com/wiki-360/resources/raw/main/Support%20Vector%20Machines%20Algorithm%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Support%20Vector%20Machines%20Algorithm%20Explained.jpg)

K-nearest neighbors (kNN) is a supervised learning algorithm that can be used to solve both classification and regression tasks. The main idea behind this algorithm is that the value or class of a data point is determined by the data points around it. kNN classifier determines the class of a data point by the majority voting principle. kNN becomes very slow as the number of data points increases because the model needs to store all data points. Thus, it is also not memory efficient. Another downside of kNN is that it is sensitive to outliers.

[![](https://github.com/wiki-360/resources/raw/main/K%20Nearest%20Neighbour%20Machine%20Learning%20Algorithms%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/K%20Nearest%20Neighbour%20Machine%20Learning%20Algorithms%20Explained.jpg)

\- To learn more in detail, Check out this [**Article**](https://lettier.github.io/posts/2016-06-10-k-nearest-neighbors-from-scratch.html)

**👉 K-Means Clustering**

K Means Clustering is one of the most popular clustering algorithms and it's the first algorithm practitioners apply when solving clustering tasks to get an idea of the structure of the dataset. It's an algorithm that, given a dataset, will identify which data points belong to each one of the k clusters. It takes your data and learns how it can be grouped. Through a series of iterations, the algorithm creates groups of data points referred to as clusters that have similar variance and that minimize a specific cost function: the within-cluster sum of squares. 

[![](https://github.com/wiki-360/resources/raw/main/K-means%20Clustering%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/K-means%20Clustering%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)

**👉 Hierarchical Clustering**

Hierarchical clustering means creating a tree of clusters by iteratively grouping or separating data points. There are two types of hierarchical clustering named Agglomerative clustering and Divisive clustering. 

Agglomerative clustering is the bottom-up approach. It merges the two points that are the most similar until all points have been merged into a single cluster. Divisive clustering is the top-down approach. It starts with all points as one cluster and splits the least similar clusters at each step until only single data points remain. One of the advantages of hierarchical clustering is that we do not have to specify the number of clusters (but we can). 

[![](https://github.com/wiki-360/resources/raw/main/Hierarchical%20Clustering%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)
](https://github.com/wiki-360/resources/raw/main/Hierarchical%20Clustering%20-%20Machine%20Learning%20Algorithms%20Explained.jpg)

**👉 Neural Networks**

Neural networks are a set of algorithms inspired by the functioning of the human brain. Generally, when you open your eyes, what you see is called data and is processed by the Neurons(data processing cells) in your brain, and recognize what is around you. That’s how similar the Neural Networks works. They take a large set of data, process the data(draws out the patterns from data), and outputs what it is. With various variants like CNN (Convolutional Neural Networks), RNN (Recurrent Neural Networks), Autoencoders, Deep Learning, etc. Neural networks are slowly becoming for data scientists or machine learning practitioners what linear regression was one for statisticians.

[![](https://github.com/wiki-360/resources/raw/main/Neural%20Networks%20Explained.png)
](https://github.com/wiki-360/resources/raw/main/Neural%20Networks%20Explained.png)

\- You can try the above shown Neural Network tool from here: [**link**](https://aegeorge42.github.io/)  

\- Neural Networks In Python 

A typical question asked by beginners, when facing various types of machine learning algorithms, is “which algorithm is the best and which one should I use?” The answer to the question varies depending on multiple factors, like (1) The size, quality, and nature of data; (2) The available computational time; (3) The urgency of the task; and (4) What you want to do with the data.

Even an experienced data scientist or machine learning engineer cannot tell which algorithm will perform the best before trying different algorithms. Although there are many other Machine Learning algorithms, these are the most popular ones and widely used. If you’re a newbie to Machine Learning, these would be a good starting point to learn. And if you are working / curious to learn more about AI, ML, Data Science, App Development, Web Development, etc, Please give a try to [**Insane**](https://play.google.com/store/apps/details?id=com.insane.android) App. A special thanks to [**Avik Jain**](https://github.com/Avik-Jain/100-Days-Of-ML-Code) for creating Infographics, [**Allison George**](https://aegeorge42.github.io/) for creating an interactive guide on neural networks from scratch, and all scientists, researchers, etc for open sourcing their codes on GitHub. 
 [https://www.theinsaneapp.com/2021/11/machine-learning-algorithms-for-beginners.html?continueFlag=3804e091429ea848e33e0bb674c06b75](https://www.theinsaneapp.com/2021/11/machine-learning-algorithms-for-beginners.html?continueFlag=3804e091429ea848e33e0bb674c06b75)
